### ISSUE TITLE
basic+ documentation of common functionality

### ISSUE THREAD

**Issue by GlenDC** [jsoverson/hackium#42]
It would be great if the common functionality and stuff used in the more common use cases are either self-documented or refer to documentation that help explain it. I think the README and CLI help documentation that's there is already a great start.

A big chunk of this project are the collection of other projects, so for anything that that uses such code directly a reference to its documentation (for that specific part) would be sufficient, sometimes a hybrid thing will be required. E.g. it would be nice to have a list of all objects exposed in the REPL, I think they already are, and for those it would be sufficient (I think) to reference the Puppeteer documentation for those object locations. Where extended it would require some extra documentation.

Important for me, and its motivation, are locations such as the interceptor where it is not very clear to me what possible values are for the ResourceType and RequestStage. For now it seems to be only possible to have `Response` as the requestStage, and I haven't seen any example not using `Script` for resource type. However for a purpose where I want a resource other than a `Script` I omit it for now, given I do not know what other resource type I can use. Omitting it seems to work as well (I guess it allows any resource type?). I tried GREPping through the code base and it doesn't seem to be something this codebase really defines, so I imagine it is another piece of software glued into here. Which is ok as well of course, if so, a reference to the documentation about those parameters would suffice as well.

And there are probably other examples that fall in the category of the examples I gave here.

I wouldn't mind contributing this myself, I would just need some pointers in the different pieces of software used and where/how they are used.

**Comment by jsoverson** [comment]
Definitely, thanks for pointing this out. I've started fleshing out documentation locally and will finish it up ASAP.

Feel free to ask questions in this issue and I'll address them as I go.

**Comment by jsoverson** [comment]
Updated the readme with more information. It doesn't add enough, but it's a start. The ways to use this project are too big for a single readme file and I need to separate them.

I'm not 100% clear on what you are referring to in your question. Those streams are from a long time ago and it may be possible with shift-refactor and shift-interpreter but I'm not 100% clear on how to answer. I just created an AMA repository for questions like this, feel free to repost there: https://github.com/jsoverson/javascript-deobfuscation-AMA

**Comment by GlenDC** [comment]
> I'm not 100% clear on what you are referring to in your question. 

It's Okay, I'm going to play with it a bit more when I get around to it. If I do not find an approach that satisfies me I'll definitely reach out. Glad to know this repo exists, I will definitely, very much so, keep it in mind :)

---

The README does look a lot better already indeed. I can understand that indeed fitting it all in the README will be hard once you expand sufficiently in your explanations. I guess your explanations related to the interceptor configuration and such are still WIP. No problem though, I'm already happy this project exists, so once agains a big thanks though!

---

I know the streams are from a long time ago, but they definitely did help a lot in understanding how I could use tools such as shift-reactor within a hackium interceptor, without those streams I was pretty lost in that part. If nothing else they explain a lot about the philosophy behind it all and how to approach these interception problems.
